server/src/main/java/io/crate/planner/optimizer/rule/MoveFilterBeneathJoin.java

```diff
@@ -67,7 +123,179 @@ public LogicalPlan apply(Filter filter,
+            return new JoinPlan(
+                getNewSource(query, join.lhs()),
+                getNewSource(query, join.rhs()),
+                join.joinType(),
+                join.joinCondition(),
+                join.isFiltered(),
+                true
+            );
+        }
+
+        Map<Set<RelationName>, Symbol> splitQueries = QuerySplitter.split(query);
+        var sources = Lists2.map(join.sources(), resolvePlan);
+        LogicalPlan lhs = sources.get(0);
+        LogicalPlan rhs = sources.get(1);
+        Set<RelationName> leftName = new HashSet<>(lhs.getRelationNames());
+        Set<RelationName> rightName = new HashSet<>(rhs.getRelationNames());
+
+        var leftAndRightQueries = divideIntoLeftAndRightQueries(splitQueries, leftName, rightName);
+        Symbol leftQuery = leftAndRightQueries[0];
+        Symbol rightQuery = leftAndRightQueries[1];
+
+        final var symbolEvaluator = new NullSymbolEvaluator(txnCtx, nodeCtx);
+        final LogicalPlan newLhs;
+        final LogicalPlan newRhs;
+        switch (join.joinType()) {
+            case LEFT:
+                /* LEFT OUTER JOIN -> NULL rows are generated for the RHS if the join-condition doesn't match
+                 *
+                 * cr> select t1.x as t1x, t2.x as t2x from t1 left join t2 on t1.x = t2.x;
+                 * +-----+------+
+                 * | t1x |  t2x |
+                 * +-----+------+
+                 * |   3 |    3 |
+                 * |   2 |    2 |
+                 * |   1 | NULL |
+                 * +-----+------+
+                 */
+                newLhs = getNewSource(leftQuery, lhs);
+                if (rightQuery == null) {
+                    newRhs = rhs;
+                } else if (couldMatchOnNull(rightQuery, symbolEvaluator)) {
+                    newRhs = rhs;
+                    splitQueries.put(rightName, rightQuery);
+                } else {
+                    newRhs = getNewSource(rightQuery, rhs);
+                }
+                break;
+            case RIGHT:
+                /* RIGHT OUTER JOIN -> NULL rows are generated for the LHS if the join-condition doesn't match
+
+                 * cr> select t1.x as t1x, t2.x as t2x from t1 right join t2 on t1.x = t2.x;
+                 * +------+-----+
+                 * |  t1x | t2x |
+                 * +------+-----+
+                 * |    3 |   3 |
+                 * |    2 |   2 |
+                 * | NULL |   4 |
+                 * +------+-----+
+                 */
+                if (leftQuery == null) {
+                    newLhs = lhs;
+                } else if (couldMatchOnNull(leftQuery, symbolEvaluator)) {
+                    newLhs = lhs;
+                    splitQueries.put(leftName, leftQuery);
+                } else {
+                    newLhs = getNewSource(leftQuery, lhs);
+                }
+                newRhs = getNewSource(rightQuery, rhs);
+                break;
+            case FULL:
```

Comment:
We can't move filters beneath full joins. We can only move a filter to the preserved side of a join assuming the filter only references columns from that side. A join side is preserved if the join returns all or a subset of the rows from
the side. If a side is not preserved, it can provide extra null rows.